documentation/modules/tracing/proc-instrumenting-producers-consumers.adoc

```diff
@@ -0,0 +1,288 @@
+// Module included in the following assemblies:
+//
+// assembly-instrumenting-kafka-clients-tracers.adoc
+
+[id='proc-instrumenting-producers-consumers-{context}']
+= Instrumenting producers and consumers for tracing
+
+[role="_abstract"]
+Use a Decorator pattern or Interceptors to instrument your Java producer and consumer application code for tracing.
+When instrumentation is added, you can record traces when messages are produced or retrieved from a topic.
+
+The OpenTracing and OpenTelemetry instrumentation projects provide classes that support instrumentation of producers and consumers.
+You enable instrumentation in producer and consumer applications by adding the tracing JARs as a dependency to your project.
+
+Decorator instrumentation:: For Decorator instrumentation, create a modified producer or consumer instance for tracing.
+Interceptor instrumentation:: For Interceptor instrumentation, add the tracing capability to the consumer or producer configuration.
+You do this by adding the `TracingProducerInterceptor` and `TracingConsumerInterceptor` instrumentation classes.
+
+.Procedure
+
+. Add the Maven dependencies for distributed tracing to the producer or consumer's `pom.xml` file.
++
+--
+.Dependencies for OpenTracing
+[source,xml,subs="attributes+"]
+----
+<dependency>
+  <groupId>io.opentracing.contrib</groupId>
+  <artifactId>opentracing-kafka-client</artifactId>
+  <version>{OpenTracingKafkaClient}</version>
+</dependency>
+----
+.Dependencies for OpenTelemetry
+[source,xml,subs="attributes+"]
+----
+<dependency>
+  <groupId>io.opentelemetry</groupId>
+  <artifactId>opentelemetry-sdk-extension-autoconfigure</artifactId>
+  <version>{OpenTelemetryVersionLower}</version>
+</dependency>
+<dependency>
+  <groupId>io.opentelemetry</groupId>
+  <artifactId>opentelemetry-semconv</artifactId>
+  <version>{OpenTelemetryVersionLower}</version>
+</dependency>
+<dependency>
+  <groupId>io.opentelemetry</groupId>
+  <artifactId>opentelemetry-sdk-trace</artifactId>
+  <version>{OpenTelemetryVersionStableLower}</version>
+</dependency>
+<dependency>
+  <groupId>io.opentelemetry.instrumentation</groupId>
+  <artifactId>opentelemetry-kafka-clients-{OpenTelemetryKafkaClient}</artifactId>
+  <version>{OpenTelemetryVersionLower}</version>
+</dependency>
+<dependency>
+  <groupId>io.opentelemetry.instrumentation</groupId>
+  <artifactId>opentelemetry-kafka-clients-common</artifactId>
+  <version>{OpenTelemetryVersionLower}</version>
+</dependency>
+<dependency>
+  <groupId>io.opentelemetry</groupId>
+  <artifactId>opentelemetry-exporter-jaeger</artifactId>
+  <version>{OpenTelemetryVersionStableLower}</version>
+</dependency>
+<dependency>
+  <groupId>io.grpc</groupId>
+  <artifactId>grpc-netty-shaded</artifactId>
+  <version>{OpenTelemetryGrpcVersionLower}</version>
+</dependency>
```

Comment:
In my experience, the following are not needed:

opentelemetry-semconv
opentelemetry-sdk-trace
opentelemetry-kafka-clients-common
grpc-netty-shaded

I have an example in my repo with producer instrumented for using the Interceptors way with following deps needed https://github.com/ppatierno/kafka-opentelemetry/blob/main/kafka-producer-auto/pom.xml#L25

I think the dependencies could be different if using the decorator and we need to differentiate the pom.xml, not having a big pom with all the dependencies for both.

Reply:
No `grpc-netty-shaded`, really? How does then gRPC work for you?

Reply:
Sorry Ales, where do you actually need gRPC for OpenTelemetry instrumenting Kafka consumer and producer. I have just run a `mvn dependency:tree` in my repo and it's not brought as an indirect dependency neither. Everything is instrumented ok and visible in Jaeger. The only thing not available yet in my repo is the usage of the "decoration".

Reply:
I added an example with "wrapping" (decorator) on my personal repo and again no gRPC is needed as well. Other than all the others dependencies above I have already listed.

Reply:
@alesj your answers are out of this discussion I will reply it anyway. Taking a look at https://github.com/open-telemetry/opentelemetry-java/tree/main/exporters/jaeger it's still gRPC but there is no need for the `grpc-netty-shaded` ... but I don't know why, because I am not gRPC expert and don't know what exactly the `grpc-netty-shaded` does.